geoh5py/ui_json/enforcers.py

```diff
@@ -0,0 +1,350 @@
+
+        self.enforcers = enforcers
+
+    @property
+    def validations(self) -> dict[str, Any]:
+        """Returns an enforcer type / validation dictionary from pool."""
+        return {k.enforcer_type: k.validations for k in self.enforcers}
+
+    def _recruit_enforcer(self, enforcer_type: str, validation: Any) -> Enforcer:
+        """
+        Create enforcer from enforcer type and validation.
+
+        :param enforcer_type: Type of enforcer to create.
+        :param validation: Enforcer validation.
+        """
+
+        if enforcer_type not in self.enforcer_types:
+            raise ValueError(f"Invalid enforcer type: {validation['type']}")
+
+        if enforcer_type == "type":
+            enforcer = TypeEnforcer(validation)
+        if enforcer_type == "value":
+            enforcer = ValueEnforcer(validation)  # type: ignore
+        if enforcer_type == "uuid":
+            enforcer = UUIDEnforcer(validation)  # type: ignore
+        if enforcer_type == "required":
+            enforcer = RequiredEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_uijson_parameters":
+            enforcer = RequiredUIJsonParameterEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_form_members":
+            enforcer = RequiredFormMemberEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_workspace_objects":
+            enforcer = RequiredWorkspaceObjectEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_object_data":
+            enforcer = RequiredObjectDataEnforcer(validation)  # type: ignore
+
+        return enforcer
+
+    def validate(self, value: Any):
+        """Validate value against all enforcers."""
+
+        for enforcer in self.enforcers:
+            self._capture_error(enforcer, value)
+
+        self._raise_errors()
+
+    def _capture_error(self, enforcer: Enforcer, value: Any):
+        """Catch and store 'BaseValidationError's for aggregation."""
+        try:
+            enforcer.enforce(self.name, value)
+        except BaseValidationError as err:
+            self._errors.append(err)
+
+    def _raise_errors(self):
+        """Raise errors if any exist, aggregate if more than one."""
+        if self._errors:
+            if len(self._errors) > 1:
+                raise AggregateValidationError(self.name, self._errors)
+            raise self._errors.pop()
+
+
+class Enforcer(ABC):
+    """
+    Base class for rule enforcers.
+
+    :param enforcer_type: Type of enforcer.
+    :param validations: Value(s) to validate parameter value against.
+    """
+
+    enforcer_type: str = ""
+
+    def __init__(self, validations: Any | None = None):
+        self._validations = validations
+
+    @abstractmethod
+    def rule(self, value: Any):
+        """True if 'value' adheres to enforcers rule."""
+        raise NotImplementedError(
+            "Enforcer is abstract. Must be implemented in subclass"

```

Comment:
Looks like a copy from below. We have so far never raised anything in geoh5py in the `abstractmethod`, as it is implied that the subclass needs to implement it. Also less to test.

Reply:
I think this is perfectly valid code.  It is sole reason that the NotImplementedError exists (https://docs.python.org/3/library/exceptions.html#NotImplementedError).  That said, I'm willing to remove for coverage, consistency reasons if you like..

Reply:
That's fair, although I have mostly seen it for the second reason (work in progress of a class). You technically can't reach that chunk of code though, since the abstract class can't be instantiated without the method being implemented.

![image](https://github.com/MiraGeoscience/geoh5py/assets/55204635/6e0d3b26-5146-4d66-b0ad-841f5aa0ca0a)

And for consistency, I would say yes, we remove.

@sebhmg Do you have an opinion generally on this.

Reply:
Just had a closer look at your example.  The point of the NotImplementedError is to prevent a developer from defining a concrete class and forgetting to implement one of the abstract methods.  So to test, we could subclass test_abstract_class, then call implement and check that the NotImplementedError is thrown.

Reply:
Did you run it? Don't think you can instantiate your subclass either without defining `message`

![image](https://github.com/MiraGeoscience/geoh5py/assets/55204635/8e3d0e1a-1e57-4fc3-86cc-eb1f8aab0518)


Seems redundant anyway, but that's just my opinion.

@sebhmg We can differ to your third-party opinion?

Reply:
`NotImplementedError` is thrown at runtime, so it will not "prevent a developer from defining a concrete class and forgetting to implement one of the abstract methods".
Form unit test maybe. However would think, as @domfournier point it out, that Python will refuse to instantiate a class with an abstract method. Only case to worry about is for `class` or `static` `abstract` methods (which looks like code small anyway). Am I missing a point? Happy to jump on a call to clarify

Reply:
and even if  it could be instantiated, python is likely to throw hat exception since it the function was declared abstract. I suspect an explicit `NotImplementedError` is useful only when not using the abstract decorator. Or when planning some features that are not supported yet. Possibly as well for a sub-class which does not implement fully the parent interface. E.g. Human::speak(), implemented on GrownUp, but not on Baby

Reply:
Looks like my test passed because the parent wasn't a pure abstract class.  You're right, if the parent is abstract, you can't instantiate without all the abstract methods, so NotImplementedError is redundant here.

Reply:
Removed